### ISSUE TITLE
Symfony4 | Expose Services

### ISSUE THREAD

**Issue by juliusstoerrle** [FriendsOfSymfony/FOSRestBundle#1790]
When using SF4 we get following ServiceNotFoundException:

> The "fos_rest.view_handler" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.

This is because in SF4 all Services are private by default.

I'm happy to submit a pull request, where I would make all services public which do not have a visibility set yet, or should i go through and only enable those ones that give me an ServiceNotFound?

As I'm a first time contributor, is there anything i should pay attention to? I did not find a contrib readme.

(Tagging @xabbuh as he just suggested me to do a PR for this)

**Comment by xabbuh** [comment]
I think we should make a case by case decision. Is each service actually meant to be used by third party code? Or is it only there for internal purposes? Services that are actually meant to be used could be turned to public to still allow users to make use of service location with the Symfony container. For the latter, we IMO don't need to do anything as users will see a deprecation with Symfony 3.4.

**Comment by juliusstoerrle** [comment]
Are there any services a consumer needs? I do not access the container directly in my code but still get the ServiceNotFound for "fos_rest.view_handler"

**Comment by xabbuh** [comment]
Does that happen when using Symfony 3.4 or 4.0? Would you be able to provide a small reproducing project?

**Comment by xabbuh** [comment]
> Are there any services a consumer needs?

I don't have any in mind. Could also be that none of them really needs to be exposed (at least there shouldn't be many that should turned into public ones).

**Comment by juliusstoerrle** [comment]
SF4, will try a demo but might take me a day or two

**Comment by xabbuh** [comment]
Could you downgrade to Symfony 3.4 and check for the exact deprecation message (there should be one)?

**Comment by juliusstoerrle** [comment]
Making a service is also needed, when the bundle itself accesses the container.
I just had a look at the [FosRestController](https://github.com/FriendsOfSymfony/FOSRestBundle/blob/master/Controller/FOSRestController.php) and that might be the culprit. Look at line 34

` $this->viewhandler = $this->container->get('fos_rest.view_handler');`

**Comment by xabbuh** [comment]
I had a look at the code and I wonder how this can happen. As far as I can see, the service seems to be set to public: https://github.com/FriendsOfSymfony/FOSRestBundle/blob/3e65ef5f5cef2ecc6f6a18d204dc5193f0feb272/DependencyInjection/FOSRestExtension.php#L246

**Comment by juliusstoerrle** [comment]
With Symfony Standard Edition 3.4-RC1 and current master I do not get any error or deprecation notice

**Comment by xabbuh** [comment]
Cool, thanks for confirming. I leave this ticket open though as we still have to look into other services we may try to retrieve from the container.

**Comment by juliusstoerrle** [comment]
I just went back to my SF4 project, and i'm still getting the ViewHandler not found error, but I can workaround that by using the ControllerTrait and a Constructor Argument `ViewHandlerInterface`.

**Comment by juliusstoerrle** [comment]
When I deserialize he is missing the `fos_rest.decoder.json`, setting `public=true `for that service in the body_listerner.xml service config, does solve the problem. I guess the same applies to `fos_rest.decoder.jsontoform` and `fos_rest.decoder.xml`, but I am not using those two.

**Comment by stof** [comment]
we should not force these services to be public, but instead use Symfony 3.3+ lazy-loading features when available instead.

**Comment by xabbuh** [comment]
I agree with @stof. See #1793 for the implementation.

**Comment by juliusstoerrle** [comment]
When throwing exceptions he does not find `fos_rest.serializer.exception_normalizer.jms`:

> 
> (2/2) ErrorExceptionWarning: call_user_func() expects parameter 1 to be a valid callback, class 'fos_rest.serializer.exception_normalizer.jms' not found
> --
> in GraphNavigator.php (line 203)

Not sure how I can  solve that one?

**Comment by xabbuh** [comment]
looks like that could be solved with schmittjoh/JMSSerializerBundle#618

**Comment by xabbuh** [comment]
Can you publish a small example project that makes it possible to reproduce it?

**Comment by juliusstoerrle** [comment]
see https://github.com/juliusstoerrle/sf4-rest-error.
My fos_rest config might be wrong too.

**Comment by xabbuh** [comment]
Thank you! I am going to check that.

**Comment by xabbuh** [comment]
@juliusstoerrle Just to give you quick feedback: I was able to reproduce the issue you reported with your application. Thanks for that. 👍 I also managed to identify one issue, but there's still some things more to fix. I'll keep you updated.

**Comment by xabbuh** [comment]
Please take a look at #1805 which will solve the first issue I was able to identify.

**Comment by juliusstoerrle** [comment]
Just pulled your changes into the example project, and there I'm getting following error: (I guess thats one of those additional issues you reported 2 days ago.)

> FatalThrowableError
> 
> Type error: Argument 1 passed to FOS\RestBundle\Serializer\JMSHandlerRegistry::__construct() must implement interface JMS\Serializer\Handler\HandlerRegistryInterface, instance of Symfony\Component\DependencyInjection\ServiceLocator given, called in C:\xampp\htdocs\my-project\var\cache\dev\ContainerGfhFFM6\getJmsSerializerService.php on line 90

As stated in my comment in the PR, I do not get any errors in my main project, even through dependencies and config are identical.

**Comment by xabbuh** [comment]
@juliusstoerrle I guess that's just related to the pending changes in schmittjoh/JMSSerializerBundle#618.

**Comment by juliusstoerrle** [comment]
Hey, when throwing exceptions I currently get:

>  The "fos_rest.serializer.exception_normalizer.jms" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead

See https://github.com/juliusstoerrle/sf4-rest-error for the code

**Comment by sanchobouillant** [comment]
@juliusstoerrle how did you fix your issue ? I have the same

**Comment by juliusstoerrle** [comment]
@sanchobouillant Do you mean the issue with the service: `fos_rest.serializer.exception_normalizer.jms`?

That is solved by the PR #1805, but as that one is still open you need to put following version constraint into your composer and run `composer update`:
`"friendsofsymfony/rest-bundle": "dev-master#dcd5add5153af7e479e97a233efa6fdcd0164879 as 2.3",`

Let me know if you need something else :)

**Comment by sanchobouillant** [comment]
that's solved my fos_rest.serializer.exception_normalizer.jms problem :D 

Do you any solution for the HandlerRegistryInterface issue ?

**Comment by deniskhmel** [comment]
Hi. How can i resolve this in my tests:
The "fos_rest.view_handler" service is private, getting it from the container is deprecated since Symfony 3.2 and will fail in 4.0. You should either make the service public, or stop using the container directly and use dependency injection instead: 3x
    2x in UserAddressControllerTest::testPostAddress from AppBundle\Tests\Controller\Api

**Comment by juliusstoerrle** [comment]
@deniskhmel What version of the bundle are you on and can you please post of `UserAddressControllerTest::testPostAddress`

**Comment by juliusstoerrle** [comment]
@sanchobouillant I think that was solved by https://github.com/schmittjoh/JMSSerializerBundle/pull/618. Are you on the latest Version of JMSSerializerBundle (^2.3 is mine). But I can't recall exactly what I did.

**Comment by deniskhmel** [comment]
yes, "jms/serializer-bundle": "^2.3",

**Comment by juliusstoerrle** [comment]
@deniskhmel  I think you might mixed up my last two comments :) 
Does your Controller extend from FosRestController? 
If so I worked around the ceprecation by creating my own base controller class:

```
<?php
declare(strict_types=1);

namespace App\Controller;

use FOS\RestBundle\Controller\ControllerTrait;
use FOS\RestBundle\View\ViewHandlerInterface;
use Symfony\Bundle\FrameworkBundle\Controller\Controller as FrameworkController;

abstract class Controller extends FrameworkController
{
    use ControllerTrait;

    /**
     * @param ViewHandlerInterface $handler
     */
    public function __construct(ViewHandlerInterface $handler)
    {
        $this->setViewHandler($handler);
    }

}
```

The Problem here is that the FosRestController wants to fetch the view handler service directly from the container, but that does not work from SF4 onward because the Service is registered as private. 

@xabbuh Where you able to track down why the service is private? (Re: https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1790#issuecomment-346330143)

**Comment by deniskhmel** [comment]
This is not controller - this is test for controller 
`
// use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Liip\FunctionalTestBundle\Test\WebTestCase;

class UserAddressControllerTest  extends WebTestCase
{
public function testGetAddresses()
    {
        $client = static::createClient();
        $client->request('GET', '/api/addresses/');
        ...
}
}
`
Before i tried to extends from: use Symfony\Bundle\FrameworkBundle\Test\WebTestCase - the same result;

**Comment by juliusstoerrle** [comment]
@deniskhmel Yes but the test is calling a controller (the one it is supposed to test). It's not the test which is the problem here.

**Comment by deniskhmel** [comment]
I made new base class controller, as you show above and extends my controller from it. Now i got error: 
{"code":500,"message":"A \"ViewHandlerInterface\" instance must be set when using the FOSRestBundle \"ControllerTrait\"."}

**Comment by juliusstoerrle** [comment]
@deniskhmel Thats unexpected, I can not see how that can happen, _ViewHandlerInterface_ is not an optional argument of the constructor.

**Comment by sanchobouillant** [comment]
@juliusstoerrle I try with 2.3 and dev-master. Still have the issue.

I will create a new issue

**Comment by XWB** [comment]
A similar issue is present in `FOSCommentBundle`:

```
The "fos_rest.view_handler" service is private, getting it from the container is deprecated since Symfony 3.2 and will fail in 4.0. You should either make the service public, or stop using the container directly and use dependency injection instead: 23x
    5x in ApiTest::testReplyToComment from FOS\CommentBundle\Tests\Functional
    4x in ApiTest::testGetCommentFlatSorted from FOS\CommentBundle\Tests\Functional
    3x in ApiTest::testAddCommentToThread from FOS\CommentBundle\Tests\Functional
    2x in ApiTest::testGetThreadFormAndSubmit from FOS\CommentBundle\Tests\Functional
    2x in ApiTest::testGetEmptyThread from FOS\CommentBundle\Tests\Functional
    2x in ApiTest::testGetCommentTree from FOS\CommentBundle\Tests\Functional
    2x in ApiTest::testGetCommentTreeDepth from FOS\CommentBundle\Tests\Functional
    2x in ApiTest::testGetCommentFlat from FOS\CommentBundle\Tests\Functional
    1x in ApiTest::testGetThread from FOS\CommentBundle\Tests\Functional
```

**Comment by andrzejpakula** [comment]
Fos rest is not compatible with symfony 4. It does not use controller as service. I have the same error. FosRestController should inherit from abstract controller and inject fos_rest_vie_handler.

**Comment by XWB** [comment]
@andrzejpakula Making the services public should do the trick.

**Comment by playmice** [comment]
What is the fix for this?

**Comment by xabbuh** [comment]
@XWB Thanks for the hint with the FOSCommentBundle builds. It allowed me to find the root cause which was now fixed by #1840.
